shell/server/pre-meteor.js

```diff
@@ -548,32 +549,70 @@ Meteor.startup(() => {
 
   WebApp.httpServer.removeAllListeners("request");
   WebApp.httpServer.on("request", (req, res) => {
-    try {
+    function nonMeteor() {
+      // Otherwise, dispatch to our own middleware proxy chain.
+      nonMeteorRequestHandler(req, res);
+      // Adjust timeouts on proxied requests to allow apps to long-poll if needed.
+      WebApp._timeoutAdjustmentRequestCallback(req, res);
+    }
+
+    Promise.resolve(undefined).then(() => {
       if (!req.headers.host) {
         res.writeHead(400, { "Content-Type": "text/plain" });
         res.end("Missing Host header");
         return;
       }
 
-      const hostname = req.headers.host.split(":")[0];
+      const hostname = req.headers.host.split(":")[0]; // strip port if it exists
       if (isSandstormShell(hostname)) {
+        if (req.url.startsWith("/_oauth/")) {
+          const parsedUrl = Url.parse(req.url, true);
+          if (parsedUrl.query && parsedUrl.query.state) {
+            const rawState = new Buffer(parsedUrl.query.state, "base64");
+            const state = JSON.parse(rawState.toString());
+            if (state.redirectUrl) {
+              const parsedRedirect = Url.parse(state.redirectUrl);
+              const redirectHostname = parsedRedirect.hostname;
+              return inMeteor(function () {
+                if (globalDb.hostIsStandalone(redirectHostname)) {
+                  res.writeHead(302, { "Location": parsedRedirect.protocol + "//" +
+                    parsedRedirect.host + req.url, });
+                  res.end();
+                  return;
+                } else {
+                  throw new Meteor.Error(400, "redirectUrl in OAuth was for an unknown host: " +
```

Comment:
You haven't handled the main domain here, so currently OAuth login doesn't work on the main domain.